File: android/src/main/kotlin/info/espeleta/flutter_tvprovider/model/TvChannel.kt
package info.espeleta.flutter_tvprovider.model

import android.graphics.Bitmap
import androidx.tvprovider.media.tv.BasePreviewProgram
import androidx.tvprovider.media.tv.PreviewChannel
import androidx.tvprovider.media.tv.PreviewProgram
import kotlinx.serialization.Serializable

/**
 * Data class representing a channel of playable programs
 */
@Serializable
data class TvChannel(
    val id: Long? = null,
    val internalProviderId: String,
    val defaultLogoUri: String,
    val logoUri: String? = null,
    val type: String? = null,
    val displayName: String? = null,
    val description: String? = null,
    val appLinkIntentUri: String? = null,
) {
    companion object {
        fun fromPreviewChannel(channel: PreviewChannel?): TvChannel {
            val providerId = if (channel?.internalProviderId == null) "" else channel.internalProviderId
            return TvChannel(
                id = channel?.id,
                internalProviderId = providerId,
                defaultLogoUri = "",
                type = channel?.type,
                displayName = channel?.displayName?.toString(),
                description = channel?.description?.toString(),
                appLinkIntentUri = channel?.appLinkIntentUri?.toString()
            )
        }
    }
}

File: lib/model/tv_channel.dart
import 'dart:typed_data';
import 'package:cbor/cbor.dart' as cbor;

/**
 * Data class representing a channel of playable programs
 */
class TvChannel {
    int? id;
    String internalProviderId;
    String defaultLogoUri;
    String? logoUri;
    String? type;
    String? displayName;
    String? description;
    String? appLinkIntentUri;

    TvChannel(
        this.internalProviderId,
        this.defaultLogoUri,
        {
            this.id,
            this.logoUri,
            this.type,
            this.displayName,
            this.description,
            this.appLinkIntentUri
        }
    );

    factory TvChannel.fromMap(Map data) {
      return TvChannel(
        data['internalProviderId'] as String,
        data['defaultLogoUri'] as String,
        id: data['id'] as int?,
        logoUri: data['logoUri'] as String?,
        type: data['type'] as String?,
        displayName: data['displayName'] as String?,
        description: data['description'] as String?,
        appLinkIntentUri: data['appLinkIntentUri'] as String?,
      );
    }

    Map<String, dynamic> toMap() {
      return <String, dynamic>{
        'id': this.id,
        'internalProviderId': this.internalProviderId,
        'defaultLogoUri': this.defaultLogoUri,
        'logoUri': this.logoUri,
        'type': this.type,
        'displayName': this.displayName,
        'description': this.description,
        'appLinkIntentUri': this.appLinkIntentUri
      };
    }

    String toString() {
      return "${this.id}: ${this.internalProviderId} - [${this.displayName}]";
    }

}